### PR TITLE
[ValSet-Pref] Cli support and tested SetValSet, DelValSet, UnDelValSet 

### PR DESCRIPTION
**Pull request by stackman27**
Part of : https://github.com/osmosis-labs/osmosis/issues/2579

## What is the purpose of the change
Add Cli tx support to tx messages. Also added test results

## Brief Changelog
n/a 

## Testing and Verifying
Manually tested the cli commands
Test hackmd: https://hackmd.io/6eb1WIyRQYiFmD1QjI-ivQ

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/valset-pref/types/msgs.go [hunk 1]
```diff
@@ -33,11 +33,6 @@ func (m MsgSetValidatorSetPreference) ValidateBasic() error {
 	totalWeight := sdk.ZeroDec()
 	validatorAddrs := []string{}
 	for _, validator := range m.Preferences {
-		_, err := sdk.ValAddressFromBech32(validator.ValOperAddress)
```
Comment (stackman27) [inline x/valset-pref/types/msgs.go:36]: Note: this check was moved to `osmoutils/cli_helpers.go`

File: x/valset-pref/client/cli/tx.go [hunk 2]
```diff
@@ -26,19 +28,42 @@ func NewSetValSetCmd() *cobra.Command {
...
+func NewUnDelValSetCmd() *cobra.Command {
+	return osmocli.TxCliDesc{
+		Use:              "undelegate-valset [delegator_addr] [amount]",
+		Short:            "UnDelegate tokens from existing valset using delegatorAddress and tokenAmount.",
+		Example:          "osmosisd tx validatorsetpreference undelegate-valset  osmo1... 100stake",
+		NumArgs:          2,
+		ParseAndBuildMsg: NewMsgUndelegateFromValidatorSet,
+	}.BuildCommandCustomFn()
+}
+
```
Comment (czarcas7ic) [inline x/valset-pref/client/cli/tx.go:56]: I dont think these need to be parsed with a custom build message (I could be wrong)

See this as an example:

https://github.com/osmosis-labs/osmosis/blob/05a9fe4262a5fcc302e98b0ddf87d9f5f538b05e/x/concentrated-liquidity/client/cli/tx.go#L25-L51

Osmocli should just be able to parse input to the message on its own

File: x/valset-pref/types/msgs.go (hunk 1 above)
Comment (czarcas7ic) [reply x/valset-pref/types/msgs.go:36]: Why did this get moved to cli_helpers? This should be checked at the validate basic level

File: x/valset-pref/types/msgs_test.go [hunk 3]
```diff
@@ -75,27 +75,6 @@ func TestMsgSetValidatorSetPreference(t *testing.T) {
...
-						Weight:         sdk.NewDecWithPrec(2, 1),
-					},
-					{
-						ValOperAddress: "osmovaloper1x2cfenmflhj3dwm2ph6nkgqr3nppkg86fxaymg",
-						Weight:         sdk.NewDecWithPrec(6, 1),
-					},
-				},
-			},
-			expectPass: false,
-		},
```
Comment (czarcas7ic) [inline x/valset-pref/types/msgs_test.go:98]: When you add this back to validate basic this should get added back as well

**Review by czarcas7ic — CHANGES_REQUESTED** [review]

File: x/valset-pref/client/cli/tx.go (hunk 2 above)
Comment (stackman27) [reply x/valset-pref/client/cli/tx.go]: That is so crazy!!! no need for parsing at all

File: x/valset-pref/types/msgs.go (hunk 1 above)
Comment (stackman27) [reply x/valset-pref/types/msgs.go]: fixed. It was just a repetitive test, i removed it from cli and added it to validation

**Comment by stackman27** [comment]
Thank you @czarcas7ic  for your feedback! implemented all the suggested changes

File: x/valset-pref/client/cli/tx.go (hunk 2 above)
Comment (czarcas7ic) [reply x/valset-pref/client/cli/tx.go]: I know, big shoutout to @ValarDragon and chatGPT lol

**Comment by czarcas7ic** [comment]
Can you ping me when tests pass?

**Comment by stackman27** [comment]
@czarcas7ic fixed test and added new weight rounding changes so that 0.999 = 1.00, 0.874 = 0.87, 0.5123 = 0.51

File: osmoutils/cli_helpers.go [hunk 4]
```diff
@@ -64,16 +64,6 @@ func ParseSdkDecFromString(s string, separator string) ([]sdk.Dec, error) {
...
 
-func ParseSdkValAddressFromString(s string, separator string) []sdk.ValAddress {
-	var parsedAddr []sdk.ValAddress
-	for _, addr := range strings.Split(s, separator) {
-		valAddr := sdk.ValAddress([]byte(addr))
-		parsedAddr = append(parsedAddr, valAddr)
-	}
-
-	return parsedAddr
-}
```
Comment (czarcas7ic) [inline osmoutils/cli_helpers.go:75]: Why was this removed?

File: x/valset-pref/client/cli/tx.go [hunk 5]
```diff
@@ -2,34 +2,52 @@ package valsetprefcli
...
 	"github.com/spf13/pflag"
 
+	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/osmosis-labs/osmosis/osmoutils"
 	"github.com/osmosis-labs/osmosis/osmoutils/osmocli"
 	"github.com/osmosis-labs/osmosis/v13/x/valset-pref/types"
 )
 
 func GetTxCmd() *cobra.Command {
 	txCmd := osmocli.TxIndexCmd(types.ModuleName)
```
Comment (czarcas7ic) [inline x/valset-pref/client/cli/tx.go:18]: Can we just change this to directly use the string valset-pref instead of utilizing validatorsetpreference for every command?

File: x/valset-pref/keeper_test.go [hunk 6]
```diff
@@ -27,15 +27,15 @@ func (suite *KeeperTestSuite) PrepareDelegateToValidatorSet() []types.ValidatorP
...
+			Weight:         sdk.NewDecWithPrec(332, 3), // 0.33
 		},
 		{
 			ValOperAddress: valAddrs[2],
 			Weight:         sdk.NewDecWithPrec(12, 2), // 0.12
 		},
 		{
 			ValOperAddress: valAddrs[3],
-			Weight:         sdk.NewDecWithPrec(348, 3), // 0.348
+			Weight:         sdk.NewDecWithPrec(348, 3), // 0.35
```
Comment (czarcas7ic) [inline x/valset-pref/keeper_test.go:38]: I didn't realize why this changed until the end. Maybe add some comment here explaining that we round after second digit in val basic

File: x/valset-pref/types/msgs.go [hunk 7]
```diff
@@ -49,9 +49,21 @@ func (m MsgSetValidatorSetPreference) ValidateBasic() error {
...
+	totalWeightFloat64, err := totalWeight.Float64()
+	if err != nil {
+		return err
+	}
+
+	roundedValueStr := fmt.Sprintf("%.2f", totalWeightFloat64)
+	roundedValue, err := strconv.ParseFloat(roundedValueStr, 64)
+	if err != nil {
+		return err
+	}
```
Comment (czarcas7ic) [inline x/valset-pref/types/msgs.go:62]: Why was this check added?

File: x/valset-pref/client/cli/tx.go (hunk 5 above)
Comment (stackman27) [reply x/valset-pref/client/cli/tx.go:18]: aah so change from `osmosisd tx validatorsetpreference...` to `osmosisd tx valset-pref` ?

File: osmoutils/cli_helpers.go (hunk 4 above)
Comment (stackman27) [reply osmoutils/cli_helpers.go:75]: I found a simpler inline way of handling this [here](https://www.youtube.com/watch?v=zBHRsYhYb-o)

also since validateBasic() checks validator address validity so this was unnecessary

File: x/valset-pref/keeper_test.go (hunk 6 above)
Comment (stackman27) [reply x/valset-pref/keeper_test.go:38]: will do

File: x/valset-pref/types/msgs.go (hunk 7 above)
Comment (stackman27) [reply x/valset-pref/types/msgs.go:62]: currently if the delegator provides {0.3333, 0.6666} as weights it'd output 0.9999 which is not equal to 1, which i think we should round up

File: x/valset-pref/client/cli/tx.go (hunk 5 above)
Comment (czarcas7ic) [reply x/valset-pref/client/cli/tx.go:18]: its not a blocker or anything, we can leave as is

**Review by czarcas7ic — APPROVED** [review]
This LGTM!

File: x/valset-pref/validator_set.go [hunk 8]
```diff
@@ -302,14 +302,38 @@ func (k Keeper) getValAddrAndVal(ctx sdk.Context, valOperAddress string) (sdk.Va
...
+		})
 	}
-	return true
+
+	return weightsRoundedValPrefList, true
+}
+
+// RoundValidatorWeights rounds to 2 digit after the decimal. For ex: 0.999 = 1.0, 0.874 = 0.87, 0.5123 = 0.51
+func (k Keeper) RoundValidatorWeights(ctx sdk.Context, totalWeight sdk.Dec) (string, error) {
+	totalWeightFloat64, err := totalWeight.Float64()
```
Comment (alexanderbez) [inline x/valset-pref/validator_set.go:330]: errr are we OK with using floats here? I'm very very weary of using arch-dependent floating point numbers in the state-machine. The `Dec` types has rounding functions that I believe do something similar to what you're trying to achieve with this method. Have you explored those?

File: x/valset-pref/types/msgs.go [hunk 9]
```diff
@@ -184,9 +196,21 @@ func (m MsgRedelegateValidatorSet) ValidateBasic() error {
 		return fmt.Errorf("The validator operator address are duplicated")
 	}
 
+	// Round to 2 digit after the decimal. For ex: 0.999 = 1.0, 0.874 = 0.87, 0.5123 = 0.51
+	totalWeightFloat64, err := totalWeight.Float64()
```
Comment (alexanderbez) [inline x/valset-pref/types/msgs.go:200]: ditto

File: x/valset-pref/validator_set.go (hunk 8 above)
Comment (stackman27) [reply x/valset-pref/validator_set.go:330]: checking right now

Comment (stackman27) [reply x/valset-pref/validator_set.go]: Replaced with SigFigRound!

**Comment by stackman27** [comment]
@alexanderbez resolved all your comments!

**Comment by stackman27** [comment]
@ValarDragon this PR is good to go!
